# Data cleaning and merging dataframes

## Loading multiple datasets

### Classical way

Reading file by file

In [222]:
import pandas as pd

path = '../data/'
orderlines = pd.read_csv(path + 'orderlines.csv')
orders = pd.read_csv(path + 'orders.csv')
brands = pd.read_csv(path + 'brands.csv')
products = pd.read_csv(path + 'products.csv')

In [223]:
products.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

In [224]:
orders.head()

order_id         created_date  total_paid      state
0    241319  2017-01-02 13:35:40       44.99  Cancelled
1    241423  2017-11-06 13:10:02      136.15  Completed
2    242832  2017-12-31 17:40:03       15.76  Completed
3    243330  2017-02-16 10:59:38       84.98  Completed
4    243784  2017-11-24 13:35:19      157.86  Cancelled

In [225]:
orderlines.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

### Another way

Using `os` and a loop to read all the files from a directory. It's also possible to read only files with a certain extension (like `.csv`):

In [226]:
import pandas as pd
import os
path = '../data/'
path, dirs, files = next(os.walk(path))
#print(files)

# remove non-csv files
for file in files:
    if not file.endswith("csv"):
        files.remove(file)

In [227]:
files

['brands.csv', 'orderlines.csv', 'orders.csv', 'products.csv']

In [228]:
# create empty list 
df_list = []

# append datasets to the list
for file in files:
    temp_df = pd.read_csv(path + file, sep=',')
    df_list.append(temp_df)

# show results
brands, orderlines, orders, products = df_list[0], df_list[1], df_list[2], df_list[3]

In [229]:
df_list[3]

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
19321  compact and portable stand vertically or horiz...  29.99     269.903   
19322  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
19321         1     12282  
19322         1      1392  
19323         1      1392  
19324         0      1392  
19325         1      1392  

[19326 rows x 7 columns]

## Data quality

### Missing values

In [230]:
def check_missing_values(data): 
    print('Missing values:' + '\n' + str(data.isna().sum()))

In [231]:
for i in range(len(df_list)): 
    print(f'File name: {files[i]}')
    check_missing_values(df_list[i])
    print('\n')

File name: brands.csv
Missing values:
short    0
long     0
dtype: int64


File name: orderlines.csv
Missing values:
id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64


File name: orders.csv
Missing values:
order_id        0
created_date    0
total_paid      5
state           0
dtype: int64


File name: products.csv
Missing values:
sku             0
name            0
desc            7
price          46
promo_price     0
in_stock        0
type           50
dtype: int64




### Duplicates

In [232]:
def check_duplicates(data): 
    print('Duplicated rows: ', data.duplicated().sum())

In [233]:
for i in range(len(df_list)): 
    print(f'File name: {files[i]}')
    check_missing_values(df_list[i])
    check_duplicates(df_list[i])
    print('\n')

File name: brands.csv
Missing values:
short    0
long     0
dtype: int64
Duplicated rows:  0


File name: orderlines.csv
Missing values:
id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64
Duplicated rows:  0


File name: orders.csv
Missing values:
order_id        0
created_date    0
total_paid      5
state           0
dtype: int64
Duplicated rows:  0


File name: products.csv
Missing values:
sku             0
name            0
desc            7
price          46
promo_price     0
in_stock        0
type           50
dtype: int64
Duplicated rows:  8746




Based on our initial exploration we need to find a way to deal with our missing values. 

Do we have other problems like data inconsistency? 

* Example: problems with the `price` column in the `orderlines` dataframe.

### Data inconsistency

In [234]:
for i in range(len(df_list)): 
    
    print(f'File name: {files[i]}')
    check_missing_values(df_list[i])
    check_duplicates(df_list[i])
    df_list[i].info()
    print('\n')

File name: brands.csv
Missing values:
short    0
long     0
dtype: int64
Duplicated rows:  0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187 entries, 0 to 186
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   short   187 non-null    object
 1   long    187 non-null    object
dtypes: object(2)
memory usage: 3.0+ KB


File name: orderlines.csv
Missing values:
id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64
Duplicated rows:  0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int

Some aspects that do not make a lot of sense:

* **products**: 
    * `price` and `promo_price` are loaded as objects. They should be floats. 
    * Quite a lot of missing values: 
        * `description`: 7 missing values. Maybe that could be inferred from the product name?
        * `price`: the missing values could be filled from the `orderlines` dataset. But first we will need to clean it. 

    
* **orders**: 
    * `created_date` should have a date data type. Then, it would be a good quality check to see if the created date for `orders` mathces with the created dates for `orderlines`. 
    
    
* **orderlines**: 
    * `unit_price` has to be a float, something wrong there. 
    * `date` has to be transformed to a date data type. Then, as we said, check that it with matches with the `orders` dataset. 
    
    
* **brands**: looks fine. 


Where do we have to start? Since `orders` and `orderlines` seem very crucial to the analysis, we will start by checking that they match.

## Cleaning orders

In [235]:
# change date datatype
orders['created_date'] = pd.to_datetime(orders['created_date'])

Some data I could extract from here to do a data quality check with `orderlines`: 

* the sum of `total_paid` on both datasets should be the same
* the initial and last date should be the same

In [236]:
min(orders['created_date'])

Timestamp('2017-01-01 00:07:19')

In [237]:
max(orders['created_date'])

Timestamp('2018-03-14 13:58:36')

In [238]:
sum(orders['total_paid'])

nan

I forgot to deal with missing values on our orders dataset!

In [239]:
orders.isna().sum()

order_id        0
created_date    0
total_paid      5
state           0
dtype: int64

#### ****Find out missing values (which values are missing)

In [240]:
# let's filter them out
orders.loc[orders['total_paid'].isna()]

order_id        created_date  total_paid    state
127701    427314 2017-11-20 18:54:39         NaN  Pending
132013    431655 2017-11-22 12:15:24         NaN  Pending
147316    447411 2017-11-27 10:32:37         NaN  Pending
148833    448966 2017-11-27 18:54:15         NaN  Pending
149434    449596 2017-11-27 21:52:08         NaN  Pending

These orders have a state of "Pending", so I can exclude them from our dataset. We are interested on the completed ones.

In [241]:
orders.dropna(inplace=True)

Now my dataset is clean. And the total paid is: 

In [242]:
orders['total_paid'].sum()

129159615.07000001

## Cleaning orderlines

First let's transform our date time. 

In [243]:
orderlines['date'] = pd.to_datetime(orderlines['date'])

In [244]:
min(orderlines['date'])
# orders: Timestamp('2017-01-01 00:07:19')

Timestamp('2017-01-01 00:07:19')

In [245]:
max(orderlines['date'])
# orders: Timestamp('2018-03-14 13:58:36')

Timestamp('2018-03-14 13:58:36')

Yes, it seems it match. Now let's test our total_paid. 

To get this value, we will have to calculate a new column, total price for each row. It would be `product_quantity` * `unit_price`. I can not multiply a two different types of data. So I will have to transform the unit price to a numeric one. 

In [246]:
orderlines.sample(3)

id  id_order  product_id  product_quantity      sku unit_price  \
164666  1442099    441424           0                 1  WAC0210      45.89   
97722   1305863    383659           0                 1  XDO0039      22.99   
208799  1508857    468517           0                 1  BOS0023      89.99   

                      date  
164666 2017-11-24 20:22:23  
97722  2017-07-29 11:10:19  
208799 2017-12-20 23:57:42

#### ****we have dots (.) in values for the unit price. FE: 1329.00 is indicated as 1.329.00

In [259]:
orderlines['product_quantity'] * pd.to_numeric(orderlines['unit_price'])

0          18.99
1         399.00
2         474.05
3          68.39
4          23.74
           ...  
293978     42.99
293979    141.58
293980     19.98
293981     19.99
293982     13.99
Length: 293983, dtype: float64

While trying to transform this column to numeric an error appears. From the error message, "Unable to parse string "1.137.99" at position 6", it seems that our dataset has some problems with the thousands separators: they were encoded as dots, and python only admits one dot per number: the _decimal_ separator!

There are a lot of different ways to solve this problem. The first thing I will do is to count how many dots appear for each number using string operations. If price has more than one dot, I will consider it corrupted. 

#### ****counting dots in unit_price column values

In [249]:
orderlines['unit_price'].str.count('\.')

0         1
1         1
2         1
3         1
4         1
         ..
293978    1
293979    1
293980    1
293981    1
293982    1
Name: unit_price, Length: 293983, dtype: int64

In [250]:
# we create a copy of the dataset
ol_temp = orderlines.copy()

In [251]:
# create a new column with the amount of dots in the unit_price column
ol_temp['dots'] = orderlines['unit_price'].str.count('\.')

In [252]:
# show the rows with more than one dot
ol_temp.query('dots > 1')

id  id_order  product_id  product_quantity      sku unit_price  \
6       1119115    299544           0                 1  APP1582   1.137.99   
11      1119126    299549           0                 1  PAC0929   2.565.99   
15      1119131    299553           0                 1  APP1854   3.278.99   
43      1119195    299582           0                 1  PAC0961   2.616.99   
59      1119214    299596           0                 1  PAC1599   2.873.99   
...         ...       ...         ...               ...      ...        ...   
293862  1649999    452946           0                 1  APP2075   2.999.00   
293887  1650045    527321           0                 1  PAC2148   3.497.00   
293889  1650050    527324           0                 1  PAC2117   3.075.00   
293911  1650088    527342           0                 1  APP2492   1.329.00   
293936  1650135    527364           0                 1  APP2491   1.329.00   

                      date  dots  
6      2017-01-01 01:17:21     2  
11     2017-01-01 02:07:42     2  
15     2017-01-01 02:14:47     2  
43     2017-01-01 08:54:00     2  
59     2017-01-01 09:53:11     2  
...                    ...   ...  
293862 2018-03-14 13:03:33     2  
293887 2018-03-14 13:10:15     2  
293889 2018-03-14 13:10:56     2  
293911 2018-03-14 13:24:51     2  
293936 2018-03-14 13:36:13     2  

[36169 rows x 8 columns]

Our theory about the thousands separators is confirmed. How can we solve this problem? 

My approach would be to remove all the dots for all the unit_prices, and then add a dot before the last 2 digits to all the rows. Then I will transform it into numeric values.

In [253]:
orderlines = orderlines.assign(unit_price_nd = orderlines['unit_price'].str.replace('\.','', regex=True))
orderlines.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                 date unit_price_nd  
0 2017-01-01 00:07:19          1899  
1 2017-01-01 00:19:45         39900  
2 2017-01-01 00:20:57         47405  
3 2017-01-01 00:51:40          6839  
4 2017-01-01 01:06:38          2374

In [254]:
orderlines['digits'] = orderlines['unit_price_nd'].str[-2:]
orderlines['integers'] = orderlines['unit_price_nd'].str[:-2]
orderlines.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                 date unit_price_nd digits integers  
0 2017-01-01 00:07:19          1899     99       18  
1 2017-01-01 00:19:45         39900     00      399  
2 2017-01-01 00:20:57         47405     05      474  
3 2017-01-01 00:51:40          6839     39       68  
4 2017-01-01 01:06:38          2374     74       23

In [255]:
orderlines['new_unit_price'] = orderlines['integers'] + '.' + orderlines['digits']
orderlines.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                 date unit_price_nd digits integers new_unit_price  
0 2017-01-01 00:07:19          1899     99       18          18.99  
1 2017-01-01 00:19:45         39900     00      399         399.00  
2 2017-01-01 00:20:57         47405     05      474         474.05  
3 2017-01-01 00:51:40          6839     39       68          68.39  
4 2017-01-01 01:06:38          2374     74       23          23.74

In [256]:
# last step is to transform it to numeric
orderlines['unit_price'] = pd.to_numeric(orderlines['new_unit_price'])
orderlines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  float64       
 6   date              293983 non-null  datetime64[ns]
 7   unit_price_nd     293983 non-null  object        
 8   digits            293983 non-null  object        
 9   integers          293983 non-null  object        
 10  new_unit_price    293983 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(4), object(5)
memory usage: 24.7+ MB


Now it is time to see if it is the same than the orders files. 

In [313]:
orderlines.drop(['unit_price_nd','digits','integers','new_unit_price'], axis=1, inplace=True)

orderlines['total_price'] = orderlines['product_quantity'] * orderlines['unit_price']
sum(orderlines['total_price'])
# 129159615.07000001

128659393.76980966

In [260]:
sum(orders['total_paid'])

129159615.06991275

What!? After all this work the values do not match! How can we figure it out where is the problem? 

## Matching `orders` and `orderlines`

It is possible that some orders exist in one dataset but not in the other one. This would be a potential source for this price mismatch. Let's solve it!

We first create a new column in the `orderlines` dataset using `assign`. We also use `isin()` to create a boolen value (True/False) that checks whether the `id_order` is present in the `orders` dataset:

In [261]:
(orderlines.
 assign(check_orders = orderlines.id_order.isin(orders.order_id))
)

id  id_order  product_id  product_quantity      sku  unit_price  \
0       1119109    299539           0                 1  OTT0133       18.99   
1       1119110    299540           0                 1  LGE0043      399.00   
2       1119111    299541           0                 1  PAR0071      474.05   
3       1119112    299542           0                 1  WDT0315       68.39   
4       1119113    299543           0                 1  JBL0104       23.74   
...         ...       ...         ...               ...      ...         ...   
293978  1650199    527398           0                 1  JBL0122       42.99   
293979  1650200    527399           0                 1  PAC0653      141.58   
293980  1650201    527400           0                 2  APP0698        9.99   
293981  1650202    527388           0                 1  BEZ0204       19.99   
293982  1650203    527401           0                 1  APP0927       13.99   

                      date unit_price_nd digits integers new_unit_price  \
0      2017-01-01 00:07:19          1899     99       18          18.99   
1      2017-01-01 00:19:45         39900     00      399         399.00   
2      2017-01-01 00:20:57         47405     05      474         474.05   
3      2017-01-01 00:51:40          6839     39       68          68.39   
4      2017-01-01 01:06:38          2374     74       23          23.74   
...                    ...           ...    ...      ...            ...   
293978 2018-03-14 13:57:25          4299     99       42          42.99   
293979 2018-03-14 13:57:34         14158     58      141         141.58   
293980 2018-03-14 13:57:41           999     99        9           9.99   
293981 2018-03-14 13:58:01          1999     99       19          19.99   
293982 2018-03-14 13:58:36          1399     99       13          13.99   

        total_price  check_orders  
0             18.99          True  
1            399.00          True  
2            474.05          True  
3             68.39          True  
4             23.74          True  
...             ...           ...  
293978        42.99          True  
293979       141.58          True  
293980        19.98          True  
293981        19.99          True  
293982        13.99          True  

[293983 rows x 13 columns]

Then, using `.query` we select rows where the value in this new column is `False`:

In [262]:
(orderlines.
 assign(check_orders = orderlines.id_order.isin(orders.order_id)).
 query("check_orders==False")
)

id  id_order  product_id  product_quantity      sku  unit_price  \
5       1119114    295310           0                10  WDT0249      231.79   
63      1119218    296284           0                 1  BNQ0042      699.00   
67      1119223    294806           0                 1  APP1849     2558.99   
69      1119226    294806           0                 1  APP1864     2797.99   
70      1119235    297261           0                 1  QNA0177      304.99   
...         ...       ...         ...               ...      ...         ...   
275665  1621177    244328           0                 1  OWC0260      349.00   
280055  1628699    261391           0                 1  APP2352     3343.00   
280352  1629247    287797           0                 1  SYN0182      484.11   
280856  1630150    261391           0                 1  SAT0091       44.99   
280869  1630178    261391           0                 1  SAT0062       49.99   

                      date unit_price_nd digits integers new_unit_price  \
5      2017-01-01 01:14:27         23179     79      231         231.79   
63     2017-01-01 09:58:35         69900     00      699         699.00   
67     2017-01-01 10:09:15        255899     99     2558        2558.99   
69     2017-01-01 10:15:14        279799     99     2797        2797.99   
70     2017-01-01 10:17:59         30499     99      304         304.99   
...                    ...           ...    ...      ...            ...   
275665 2018-02-20 13:00:45         34900     00      349         349.00   
280055 2018-02-27 11:43:25        334300     00     3343        3343.00   
280352 2018-02-27 19:16:26         48411     11      484         484.11   
280856 2018-02-28 12:18:26          4499     99       44          44.99   
280869 2018-02-28 12:30:57          4999     99       49          49.99   

        total_price  check_orders  
5           2317.90         False  
63           699.00         False  
67          2558.99         False  
69          2797.99         False  
70           304.99         False  
...             ...           ...  
275665       349.00         False  
280055      3343.00         False  
280352       484.11         False  
280856        44.99         False  
280869        49.99         False  

[240 rows x 13 columns]

It looks like 240 rows in `orderlines` come from orders not present in the `orders` dataset. This is quite inconsistent, since the `orders` dataset should be the one and only source of truth for orders: if an order is not there, it should not exist. We will try to find out why that happened, but for now, let's just remove those "ghost" orders:

In [263]:
orderlines = (orderlines.
              assign(check_orders = orderlines.id_order.isin(orders.order_id)).
              query("check_orders==True")
             ).copy()

Now let's look at this problem in the opposite direction: are there orders in the `orders` dataset not prsent in `orderlines`?

In [264]:
(orders.
 assign(check_orders = orders.order_id.isin(orderlines.id_order)).
 query("check_orders==False")
)

order_id        created_date  total_paid            state  \
8         245941 2017-01-01 10:32:23      183.52        Completed   
65        268629 2017-01-31 11:27:25       73.23        Completed   
75        272862 2017-04-18 18:17:10      128.99          Pending   
82        277655 2017-02-01 08:44:55        9.99          Pending   
83        277994 2017-01-23 18:30:11       52.99        Completed   
...          ...                 ...         ...              ...   
226835    527328 2018-03-14 13:14:12        0.00  Shopping Basket   
226851    527344 2018-03-14 13:25:21        0.00      Place Order   
226853    527346 2018-03-14 13:26:14        0.00      Place Order   
226855    527348 2018-03-14 13:28:18        0.00      Place Order   
226889    527382 2018-03-14 13:47:43        0.00  Shopping Basket   

        check_orders  
8              False  
65             False  
75             False  
82             False  
83             False  
...              ...  
226835         False  
226851         False  
226853         False  
226855         False  
226889         False  

[22213 rows x 5 columns]

There are more than 22000 orders in the `orders` dataset that are not present on the `orderlines` dataset!!! We can try to find out why by looking at the state of these orders:

In [265]:
(orders.
 assign(check_orders = orders.order_id.isin(orderlines.id_order)).
 query("check_orders==False").
 state.value_counts()
)

Place Order        12304
Shopping Basket     9810
Completed             45
Cancelled             41
Pending               13
Name: state, dtype: int64

It looks like most of them are orders that were not fully completed: the products were left in the shopping basket or the order was "placed" but maybe not paid (hence the state "Place Order". Some of them were "Completed", though. 

This will require further research, and we might have to come back to these orders if we have to explore consumer behaviour (e.g. why are orders left in the shopping basket?), but for now, for the sake of data quality, let's drop all of these unmatched orders:

In [266]:
orders = (orders.
          assign(check_orders = orders.order_id.isin(orderlines.id_order)).
          query("check_orders==True")
         ).copy()

Let's now check again if the total paid matches:

In [267]:
orders.total_paid.sum()

129130877.18

In [268]:
orderlines.total_price.sum()

128659393.77

STILL NOT MATCHING!!! This is outrageous. Let's keep exploring.

## Solving the price mismatch

Let's merge both datasets and compare, order by order, the `total_price`. We will call this new merged dataset `orders_info`.

*Note: To merge `orderlines` with `orders`, we have to group and aggregate `orderlines`!

In [269]:
orders_info = (
orderlines
    .groupby('id_order')
    .agg({'total_price':'sum'})
    .merge(orders, how='left', left_on='id_order', right_on='order_id')
    .copy()
)
orders_info

total_price  order_id        created_date  total_paid  \
0             44.99    241319 2017-01-02 13:35:40       44.99   
1            129.16    241423 2017-11-06 13:10:02      136.15   
2             10.77    242832 2017-12-31 17:40:03       15.76   
3             77.99    243330 2017-02-16 10:59:38       84.98   
4            153.87    243784 2017-11-24 13:35:19      157.86   
...             ...       ...                 ...         ...   
204686        42.99    527397 2018-03-14 13:56:38       42.99   
204687        42.99    527398 2018-03-14 13:57:25       42.99   
204688       141.58    527399 2018-03-14 13:57:34      141.58   
204689        19.98    527400 2018-03-14 13:57:41       19.98   
204690        13.99    527401 2018-03-14 13:58:36       18.98   

                  state  check_orders  
0             Cancelled          True  
1             Completed          True  
2             Completed          True  
3             Completed          True  
4             Cancelled          True  
...                 ...           ...  
204686      Place Order          True  
204687  Shopping Basket          True  
204688  Shopping Basket          True  
204689  Shopping Basket          True  
204690      Place Order          True  

[204691 rows x 6 columns]

Now that the `total` from both datasets is in the same dataframe, we can create a new column with the difference:

In [270]:
orders_info['price_difference'] = orders_info['total_paid'] - orders_info['total_price']
orders_info.query('state == "Completed"').sample(10)

total_price  order_id        created_date  total_paid      state  \
138230       467.99    452727 2017-11-30 16:45:48      467.99  Completed   
159718        87.99    476658 2017-12-29 19:00:49       91.98  Completed   
75754        398.97    382356 2017-07-26 15:53:54      398.97  Completed   
132923       549.00    446865 2017-11-27 03:28:26      549.00  Completed   
4082          44.99    303937 2017-01-09 00:44:26       56.98  Completed   
66645        879.00    372034 2017-07-06 11:10:20      879.00  Completed   
193261       322.24    514589 2018-02-19 23:31:13      326.23  Completed   
167601         6.99    485645 2018-01-08 17:38:05       13.98  Completed   
130713       301.93    444440 2017-11-26 01:12:30      308.92  Completed   
163383       259.99    480855 2018-01-03 13:10:03      259.99  Completed   

        check_orders  price_difference  
138230          True              0.00  
159718          True              3.99  
75754           True              0.00  
132923          True              0.00  
4082            True             11.99  
66645           True              0.00  
193261          True              3.99  
167601          True              6.99  
130713          True              6.99  
163383          True              0.00

In [271]:
orders_info.price_difference.describe()

count    204691.000000
mean          2.303391
std         297.461827
min      -36159.900000
25%           0.000000
50%           0.000000
75%           0.010000
max       90898.610000
Name: price_difference, dtype: float64

In [272]:
temp = (
orders_info
    .assign(price_diff = round(orders_info['total_paid'] - orders_info['total_price'], 2))
    ['price_diff']
    .unique()
    .tolist()
)

pd.DataFrame({
    'x':temp
}).sort_values(by='x').head(10)

x
97  -36159.90
210  -7740.01
168  -3106.00
208  -3024.81
206  -2335.41
211  -1961.41
153  -1931.98
161  -1554.19
166  -1122.67
209   -381.47

Looks like the maximum and minimun price differences are huge: some orders are really corrupted. But we also see from the mean and the quartiles that the price difference is neglegible for most orders.

## Challenge: Remove outliers

Decide on a criteria for removing orders whenever you cannot trust the price difference between `orders` and `orderlines`. 

In [ ]:
# code here

Save the data once you are sure you can trust it!

In [303]:
orderlines.to_csv(path + 'orderlines_cl.csv', index=False)
orders.to_csv(path + 'orders_cl.csv', index=False)

## Challenge: Cleaning products

Now it is time to clean the products dataset. Let's do a quick review of the things where wrong: 

In [ ]:
print(products.info())
check_duplicates(products)
check_missing_values(products)

Looking at this overview, we can see that there are different things that has to be changed: 

* Data types: 
    * price should be a float
    * promo price should be a float
* Duplicated rows. They have to be removed. 
    * To accomplish this step you can use the method `pd.DataFrame.drop_duplicates()`. Be sure you drop all the duplicates based on the column **sku**, as it is the one you will use to merge with orderlines. 
* Missing values: 
    * Description may be can be infered by the name
    * price. Is there a way we can extract the information from another table?
    * type. Do we need this column for our analysis?
    
This task can be accomoplished by using all the methods you already know.

### Start of the challenge

### Duplicates

In [274]:
products.drop_duplicates(inplace=True)

In [275]:
products.shape

(10580, 7)

Check if you have a unique description for each sku:

In [276]:
products.groupby('sku')['sku'].count().sort_values(ascending=False).head(5)

sku
APP1197    2
PAC1555    1
PAC1557    1
PAC1558    1
PAC1559    1
Name: sku, dtype: int64

The product APP1197 is not unique. Let's zoom in:

In [277]:
products.query('sku == "APP1197"')

sku                                               name  \
7992  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   
8000  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   

                                                   desc price promo_price  \
7992  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...  1729     1305.59   
8000  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...   NaN     1305.59   

      in_stock  type  
7992         0  1282  
8000         0  1282

In [278]:
products.drop_duplicates('sku').query('sku == "APP1197"')

sku                                               name  \
7992  APP1197  Apple iMac 21.5 "Core i5 31 GHz Retina display...   

                                                   desc price promo_price  \
7992  Desktop Apple iMac 21.5 inch i5 31 GHz Retina ...  1729     1305.59   

      in_stock  type  
7992         0  1282

I can drop the duplicates only focusing on the column 'sku':

In [279]:
products.drop_duplicates('sku', inplace=True)

#### Data types

In [ ]:
# pd.to_numeric(products['price'])

It seems this time that the problem in our products prices is different than the one we have found on orderlines. How is it possible we have more than 3 digits?

Let's look at a random sample in our dataframe to understant more about it:

In [280]:
products.sample(10)

sku                                               name  \
15723    SNN0057              Sonnet Thunderbolt Cable 0.5m Black 3   
17418    AP20308              Like new - Apple iPhone SE 128GB Gold   
10736    ICA0055   Incase Systm Smart Case Cover iPhone 6 / 6S Rosa   
14636    WDT0369  My Passport for Mac 3TB External Hard Drive US...   
483    OWC0054-2     Mac OWC memory 4GB (2x2GB) 667MHz DDR2 SO-DIMM   
13191    PAC1691  Pack QNAP TS-251A NAS Server | 8GB RAM | Seaga...   
15666  OTT0137-A  Open - Otterbox Clearly Protected gel Case iPh...   
11518  WDT0243-A  Open - Western Digital Network PC and Mac 6TB ...   
77       WAC0046                     Intuos Wacom stylus ArtPen 5/4   
2285     BTC0016  Boostcase Hybrid Case & 2700mAh External Batte...   

                                                    desc       price  \
15723  Cable length 05 m with connection Thunderbolt ...       28.44   
17418          iPhone SE reconditioned 4 inch Free 128GB         529   
10736                 Double casing layer iPhone 6 / 6s.       34.95   
14636  3TB external hard drive compatible with Apple ...      169.99   
483    RAM 4GB (2x2GB) Mac mini and iMac (2006/07) Ma...       47.98   
13191  NAS with 8GB RAM and 16TB (2x8TB) Seagate Iron...  11.356.489   
15666  transparent cover and a flexible piece TPU for...       19.99   
11518     WD hard drive designed for NAS 6TB (WD60EFRX).         255   
77     special stylus Mac and PC for Intuos4 graphics...       109.9   
2285   Protective housing with built 2700mAh battery ...       99.99   

      promo_price  in_stock      type  
15723     239.895         0      1325  
17418   3.989.975         0  51601716  
10736     219.905         0  11865403  
14636     130.795         0  11935397  
483       419.894         1      1364  
13191   9.611.792         0  12175397  
15666      88.366         0      1298  
11518   1.821.965         0  12655397  
77        999.944         0      1229  
2285      399.905         0  5,49E+11

My strategy would be:

    1. Replace missing values with a "special" string. In that case I will put "000.000" as it would represent that this value has a wrong format and has to be changed. This step is done to avoid future problems while cleaning the column price.
    2. Create a new column called "price_split" with the values of price splited by the ","
    3. Create a new column called "dots_count" counting the number of dots a price contains

In [281]:
import numpy as np

products = (
products
    # Replace all missing values by '000.000'
    .assign(price = lambda x: x['price'].replace(np.nan, '000.000'))
    .assign(price_split = lambda x: x['price'].str.split('\.'),
            dots_count = lambda x: x['price'].str.count('\.'))
)
products.head(10)

sku                                               name  \
0    RAI0007                  Silver Rain Design mStand Support   
1    APP0023                  Apple Mac Keyboard Keypad Spanish   
2    APP0025                   Mighty Mouse Apple Mouse for Mac   
3    APP0072      Apple Dock to USB Cable iPhone and iPod white   
4    KIN0007        Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
5    APP0073     Apple Composite AV Cable iPhone and iPod white   
6    KIN0008        Mac Memory Kingston 1GB 667MHz DDR2 SO-DIMM   
7    KIN0009        Mac Memory Kingston 2GB 800MHz DDR2 SO-DIMM   
8  KIN0001-2  Mac memory Kingston 4GB (2x2GB) 667MHz DDR2 SO...   
9    APP0100             Apple Adapter Mini Display Port to VGA   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
5  IPhone and iPod AV Cable Dock to Composite Video.     45     420.003   
6  1GB RAM Mac mini and iMac (2006/07) MacBook Pr...  18.99     146.471   
7       2GB RAM iMac with Intel Core 2 Duo (Penryn).  36.99     274.694   
8  RAM 4GB (2x2GB) Mac mini and iMac (2006/07) Ma...     74     669.904   
9  Adapter Mini Display Port to VGA MacBook and M...     35     330.003   

   in_stock      type price_split  dots_count  
0         1      8696    [59, 99]           1  
1         0  13855401        [59]           0  
2         0      1387        [59]           0  
3         0      1230        [25]           0  
4         1      1364    [34, 99]           1  
5         0      1230        [45]           0  
6         0      1364    [18, 99]           1  
7         0      1364    [36, 99]           1  
8         0      1364        [74]           0  
9         0      1325        [35]           0

Now I need to look at each value in the new column "price_split" and check the folliwing condition:

If the last element of the value in price split has a length bigger than 2 then add a True inside the list need_check, else add a False.

Example 1:

Input: price_split = [[59],[99]] The last element of the value in price split is [99]. The length of this element is 2, as it only have 2 characters. Then we will append a False on the list need_check.

Example 2:

Input: price_split = [[234],[895]] The last element of the value in price split is [895]. The length of the last element is 3, as it has 3 characters. Then we will append a True on the list need_check.

In [282]:
need_check = []
len_split = []
for val in products['price_split']:    
    len_split.append(len(val))
    if len(val[-1]) > 2: 
        need_check.append(True)
    else:
        need_check.append(False)

We have created two lists:

* need_check: tells us which element of the dataframe has a last element on price_split with a lenght higher than 2.
* len_split : tells us how many elements are inside each price_split.
     * Example 1: [[59],[99]] it has a lenght of 2
     * Example 2: [59] it has a lenght of 1
     * Example 3: [[654],[998]] it has a lenght of 2
     * Example 4: [[9],[654],[998]] it has a lenght of 3

Then we can add this elements into our products dataframe to detect these rows:

In [283]:
(
products
    .assign(need_check = need_check, 
            len_split = len_split)
    .query('need_check==True & len_split > 1 & dots_count != 0')
).head()

sku                                               name  \
34   TWS0019  Twelve South MagicWand support Apple Magic Tra...   
362  REP0043                      Speaker lower repair iPhone 4   
480  PIE0011                     Internal Battery for iPhone 3G   
515  SEN0061  Sennheiser EZX 80 Handsfree iPhone iPad and iP...   
518  SEV0026  Service installation RAM + HDD + SSD MacBook /...   

                                                  desc    price promo_price  \
34   MagicWand for wireless keyboard and Magic Trac...  000.000     299.899   
362  Repair service including parts and labor for i...  499.004     499.004   
480        Replacement AC Adapter for Apple iPhone 3G.   98.978      98.978   
515  IPhone bluetooth headset with microphone iPad ...  649.891     649.891   
518  RAM + HDD installation + SSD in your MacBook /...  599.918     599.918   

     in_stock      type price_split  dots_count  need_check  len_split  
34          0      8696  [000, 000]           1        True          2  
362         0  1,44E+11  [499, 004]           1        True          2  
480         0  21485407   [98, 978]           1        True          2  
515         0      5384  [649, 891]           1        True          2  
518         0  20642062  [599, 918]           1        True          2

By defining a set of rules, I can quickly detect the prices that are giving me errors. Those rules are:

* The length of the last element on the value price_split has a lenght bigger than 2 (This is what the column need_check is telling us).
* There is more than 1 element inside the column price_split (Information given by the column len_split).
* The number of dots in price is different than 0 (Column dots_count).

The next steps would be to use this conditions to replace this values by missing values. To do it, I am going to use the function from the module numpy called np.where. It works as an if else statement; if a condition is accomplished, then return a value, else, return another.

In [285]:
products = (
products
    .assign(need_check = need_check, 
           len_split = len_split)
    .assign(
        price = lambda x: pd.to_numeric(
            np.where((x['need_check']==True) & (x['len_split'] > 1) & (x['dots_count'] !=0), np.nan, x['price'])
        )
    )
)
products.query('need_check==True & len_split > 1 & dots_count !=0').sample(10)

sku                                               name  \
15902    PAC2006  Synology DS1517 + | 8GB RAM | 40TB (5x8TB) Sea...   
11261    PAC1678          QNAP TS-453A | 16GB | 32TB (4x8TB) WD Red   
15379    REP0326                Headset calls iPhone 6s Plus repair   
1059     MOP0059  Mophie Space Pack Battery Case (1700mAh) and S...   
15387    REP0287                 iPhone repair posterior chamber 5c   
10742  ADN0023-A            Open - Adonit Jot Stylus Black Mini 2.0   
19251    LIN0014          Linksys Wi-Fi Velop system AC4400 2 units   
13109    TPL0029  TP-LINK TL-WPA8630 3-port Gigabit Powerline Ki...   
15840  APP1565-A           Open - Apple Lightning to USB Cable 1m-C   
2003     PAC0716  Crucial MX200 expansion kit 1TB + 8GB 1066Mhz ...   

                                                    desc  price promo_price  \
15902           NAS with 8GB RAM and 40TB for Mac and PC    NaN  20.069.629   
11261  NAS TS-453A with 4GB of memory RAM + 32TB (4x8...    NaN  20.633.682   
15379  Repair service including parts and labor for i...    NaN       69.99   
1059   Housing with battery and 32GB external storage...    NaN   1.599.862   
15387  Repair service including parts and labor for i...    NaN     599.906   
10742                  Mini Pointer for iPhone and iPad.    NaN     168.126   
19251     Wi-Fi high-density intelligent Mesh technology    NaN   2.999.905   
13109  Network Extender by Lonea Electric PLC av1200 ...    NaN   1.219.898   
15840  USB-C connection cable 1m Lightning for iPhone...    NaN     248.176   
2003   expansion 8GB (2x4GB) 1066MHZ + Crucial MX200 ...    NaN   7.169.903   

       in_stock      type     price_split  dots_count  need_check  len_split  
15902         0  12175397  [27, 187, 885]           2        True          3  
11261         0  12175397  [23, 932, 382]           2        True          3  
15379         0  1,44E+11   [6, 999, 003]           2        True          3  
1059          0  5,49E+11   [1, 599, 862]           2        True          3  
15387         0  1,44E+11      [599, 906]           1        True          2  
10742         0      1298      [199, 904]           1        True          2  
19251         1      1334   [2, 999, 905]           2        True          3  
13109         0      1334      [181, 379]           1        True          2  
15840         0      1298      [290, 001]           1        True          2  
2003          0      1433   [7, 629, 597]           2        True          3

As you can see, we replaced all the values with wrong prices with missing data. New let's drop all the columns we do not need:

In [286]:
products.drop(['price_split','need_check','len_split','dots_count'], axis=1, inplace=True)
check_missing_values(products)

Missing values:
sku              0
name             0
desc             7
price          587
promo_price      0
in_stock         0
type            50
dtype: int64


And now is time to find a strategy to fill this missing values.

In [287]:
products.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10579 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          10579 non-null  object 
 1   name         10579 non-null  object 
 2   desc         10572 non-null  object 
 3   price        9992 non-null   float64
 4   promo_price  10579 non-null  object 
 5   in_stock     10579 non-null  int64  
 6   type         10529 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 661.2+ KB


#### Missing values

As now we have incremented the number of missing data (we transformed all the wrong prices to missing ones) let's find out an strategy to fill them.

But first, let's look if they are important in our dataset:

In [288]:
prod_na = products.loc[products.isna().sum(axis=1) > 0,:].copy()
prod_na

sku                                               name  \
34     TWS0019  Twelve South MagicWand support Apple Magic Tra...   
307    SAN0017       SanDisk Cruzer Edge USB 2.0 Flash Drive 16GB   
362    REP0043                      Speaker lower repair iPhone 4   
480    PIE0011                     Internal Battery for iPhone 3G   
515    SEN0061  Sennheiser EZX 80 Handsfree iPhone iPad and iP...   
...        ...                                                ...   
19312  REP0424                       Input repair Headphones iPad   
19313  REP0421                     iPad charging connector repair   
19314  REP0416                           iPad front camera repair   
19315  REP0413                            repair rear camera iPad   
19316  REP0407                                iPad battery repair   

                                                    desc  price promo_price  \
34     MagicWand for wireless keyboard and Magic Trac...    NaN     299.899   
307                         Pendrive USB 2.0 Mac and PC.    8.0      59.895   
362    Repair service including parts and labor for i...    NaN     499.004   
480          Replacement AC Adapter for Apple iPhone 3G.    NaN      98.978   
515    IPhone bluetooth headset with microphone iPad ...    NaN     649.891   
...                                                  ...    ...         ...   
19312  Repair service including parts and labor for iPad    NaN       69.99   
19313  Repair service including parts and labor for iPad    NaN       69.99   
19314  Repair service including parts and labor for iPad    NaN       69.99   
19315  Repair service including parts and labor for iPad    NaN       69.99   
19316  Repair service including parts and labor for iPad    NaN       69.99   

       in_stock      type  
34            0      8696  
307           0       NaN  
362           0  1,44E+11  
480           0  21485407  
515           0      5384  
...         ...       ...  
19312         0  1,44E+11  
19313         0  1,44E+11  
19314         0  1,44E+11  
19315         0  1,44E+11  
19316         0  1,44E+11  

[639 rows x 7 columns]

Decide which percentage of the data represents. Always think smart about using your time:

* Which % of the orders contains one of these products? In that case I has to look at orderlines.

In [289]:
orderlines.shape[0]

293743

In [293]:
# percentage of orders with products with missing prices on products dataset
orderlines.loc[orderlines['sku'].isin(prod_na['sku']),:].shape[0] / orderlines.shape[0]

0.025604014393534484

In [294]:
# this is interesting, as for the products data the percentage is much smaller
prod_na.shape[0] / products.shape[0]

0.06040268456375839

Which impact has on the total revenue?

In [295]:
orderlines.loc[orderlines['sku'].isin(prod_na['sku']),:]['total_price'].sum() / orderlines['total_price'].sum()

0.027227081578374515

With a 3% only on the total revenue, but we are considering all the orders, and we are interested on the completed ones.

In [297]:
na_skus = prod_na['sku'].tolist()

(
orderlines
    .merge(orders, how='left', left_on='id_order', right_on='order_id')
    .query('state == "Completed" & sku == @na_skus')
    ['total_price'].sum()
) / (
orderlines
    .merge(orders, how='left', left_on='id_order', right_on='order_id')
    .query('state == "Completed"')
    ['total_price'].sum()
)

0.02922734355997912

As you can see, the products that are missing only represents 2.9% of the total invoice. I will not drop them, as it still has some information which I would like to keep it to create categories (sku, item name and item description), so I am going to save it with prices as missing values. But I will add an extra column telling me which is the maximum price for these skus in orderlines.

In [298]:
# find our which sku are missing in products
prod_na_list = (
products
    .loc[products['price'].isna(),'sku']
    .tolist()
)

# find them on orderlines, and aggregate the information to find the maximum value
orderlines_prod_na_max = (
orderlines
    .query('sku == @prod_na_list')
    .groupby('sku')
    .agg({'unit_price':'max'})
    .rename(columns={'unit_price':'max_price_orderlines'})
    .reset_index()
)
orderlines_prod_na_max[:5]

sku  max_price_orderlines
0  ADN0023-A                 16.81
1  APP0660-A                253.11
2    APP0869                539.00
3    APP0870                539.00
4    APP0871                539.00

Time to add the column max_price_orderlines in our products data:

In [299]:
products = (
products
    .merge(orderlines_prod_na_max, how='left',on='sku')
)
products.loc[products['max_price_orderlines'].notna()].sample(10)

sku                                               name  \
2981    IOT0018         iOttie iTap magnetic car holder for iPhone   
7007    REP0337                        Repair Speaker lower iPad 3   
6473    PAC1902  SSD Crucial MX300 expansion kit 1TB + 8GB 1600...   
6938    REP0304                   Rear Camera Repair iPhone 6 Plus   
6912    REP0265                    Load Connector Repair iPhone 4s   
6937    REP0287                 iPhone repair posterior chamber 5c   
6927    REP0306                 Headset calls iPhone 6 Plus Repair   
6828  QNA0198-A                            Open - QNAP TS-231P NAS   
368     SEV0027            Budget Request repair Apple MacBook Air   
378     SEV0030                   Apple iMac repair Budget Request   

                                                   desc  price promo_price  \
2981  Magnetic car holder compatible with all models...    NaN     249.901   
7007  Repair service including parts and labor for i...    NaN       69.99   
6473  SSD expansion 1TB + 8GB 1600Mhz RAM for Mac mi...    NaN     384.584   
6938  Repair service including parts and labor for i...    NaN       69.99   
6912  Repair service including parts and labor for i...    NaN     399.905   
6937  Repair service including parts and labor for i...    NaN     599.906   
6927  Repair service including parts and labor for i...    NaN     599.906   
6828  2-bay NAS server with 2 LAN ports and a powerf...    NaN   1.869.813   
368                    diagnosis to repair MacBook Air.    NaN     199.892   
378                           diagnosis to repair iMac.    NaN     199.892   

      in_stock      type  max_price_orderlines  
2981         1      5720                 24.99  
7007         0  1,44E+11                 69.99  
6473         1      1433                386.95  
6938         0  1,44E+11                 69.99  
6912         0  1,44E+11                 39.99  
6937         0  1,44E+11                 59.99  
6927         0  1,44E+11                 59.99  
6828         0      1298                186.98  
368          0  1,46E+11                 19.99  
378          0  1,46E+11                 19.99

And not let's store it as a products clean csv into our path:

In [301]:
products.drop(['promo_price','in_stock','type'], axis=1).to_csv(path + 'products_cl.csv', index=False)

## Brands

The brands csv looks fine, so we can work together with it. 

In [302]:
brands.to_csv(path + 'brands_cl.csv', index=False)